workflows/diagnostics/fv3net/diagnostics/reservoir/compute.py

```diff
@@ -0,0 +1,184 @@
+import argparse
+import fsspec
+import numpy as np
+import os
+from tempfile import NamedTemporaryFile
+from typing import Union, Optional, Sequence
+import xarray as xr
+import vcm
+import yaml
+
+import fv3fit
+from fv3fit.reservoir.utils import get_ordered_X
+from fv3fit.reservoir import ReservoirComputingModel, HybridReservoirComputingModel
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+ReservoirModel = Union[ReservoirComputingModel, HybridReservoirComputingModel]
+
+
+def _get_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("reservoir_model_path", type=str, help="Input Zarr path")
+    parser.add_argument("output_path", type=str, help="Directory to save outputs to")
+    parser.add_argument(
+        "validation_config_path", type=str, help="Path to validation data config"
+    )
+    parser.add_argument(
+        "n_synchronize",
+        type=int,
+        help=(
+            "Number of timesteps from start of validation to use in reservoir "
+            "synchronization (not used in prediction)."
+        ),
+    )
+    parser.add_argument(
+        "--n-validation-batches",
+        type=int,
+        default=None,
+        help="Number of batch data netcdfs to use for validation. Defaults to use all.",
+    )
+    return parser
+
+
+def _load_batches(path, variables, nfiles):
+    ts_loader = fv3fit.data.NCDirLoader(
+        url=path,
+        nfiles=nfiles,
+        dim_order=["time", "x", "y", "z"],
+        varying_first_dim=True,
+        sort_files=True,
+        shuffle=False,
+    )
+    tfdata = ts_loader.open_tfdataset(
+        variable_names=variables, local_download_path=None
+    )
+    return tfdata
+
+
+def _get_variables_to_load(model: ReservoirModel):
+    variables = list(model.input_variables)
+    if isinstance(model, HybridReservoirComputingModel):
+        return variables + list(model.hybrid_variables)
+    else:
+        return variables
+
+
+def _get_predictions_over_batch(
+    model: ReservoirModel,
+    states_with_overlap_time_series: Sequence[np.ndarray],
+    hybrid_inputs_time_series: Optional[Sequence[np.ndarray]] = None,
+):
+    prediction_time_series = []
+    n_timesteps = states_with_overlap_time_series[0].shape[0]
+    for t in range(n_timesteps):
+        state = [
+            variable_time_series[t]
+            for variable_time_series in states_with_overlap_time_series
+        ]
+        model.increment_state(state)
+        predict_kwargs = {}
+        if hybrid_inputs_time_series is not None:
+            predict_kwargs["hybrid_input"] = [
```

Comment:
Should this be t+1 for the hybrid variables?

Reply:
Or perhaps the hybrid variables saved are the post t+1 data already? (For a 15 minute fv3gfs step)

Reply:
You're right- that's a subtle but important point for the imperfect 'prediction' to also be for the same t+1 timestep as the reservoir prediction. It's a bug in the training too. I also fixed that line in the training in the review response.

Reply:
Just refreshed myself on the [training data generation](https://github.com/ai2cm/vcm-workflow-control/blob/2023-06-23-reservoir-workflow-example/examples/generate_hybrid_training/prescribe_reference_3h.yaml) to double check your second point- both the prescribed state (target) and the hybrid input (ML prediction) will both be for the same end of timestep as long as the state is saved to state_after_timestep.zarr and the prescribing is done via the bias_correction config field (naming is a relic from when the feature was introduced, should probably update that)

Reply:
On second thought- 
You are correct that it should be the ML model prediction at time t+1 used as input, but in correcting this I'm thinking that it might be best to take care of this shift in the data preprocessing. e.g. a timestep in the data might have timestamp 0100, with data air_temperature (at 0100) and air_temperature_predicted_by_corrective_ML_subsequent_timestep (prediction for 0400).

The advantage of this would be that 
i) all the data needed for a prediction would be contained in a single batch, no need to grab the first ML-prediction from the next batch when at the end of a batch
ii) allows for hybrid prediction cases like Paula's where the additional inputs aren't necessarily an imperfect prediction for the predicted timestep, but are just other fields at the current input state time